federation/pkg/federatedtypes/scheduling.go

```diff
@@ -87,6 +93,57 @@ func (a *replicaSchedulingAdapter) IsSchedulingAdapter() bool {
 	return true
 }
 
+func isSelected(names []string, name string) bool {
+	for _, val := range names {
+		if val == name {
+			return true
+		}
+	}
+	return false
+}
+
+func isObjHpaControlled(fedObj pkgruntime.Object) (bool, error) {
+	hpaSelectedClusters, error := hpautil.GetHpaTargetClusterList(fedObj)
+	if error != nil {
+		return false, error
+	}
+
+	if hpaSelectedClusters == nil {
+		return false, nil
+	}
+	return true, nil
+}
+
+// initializeScheduleState initializes the schedule state for consumption by schedule
+// functions (schedule or simple schedule). After this initialization the state would
+// already have information, if only a subset of clusters targetted by hpa, or all clusters
+// need to be considered by the actual scheduling functions.
+func initializeScheduleState(fedObj pkgruntime.Object, clusterNames []string) (map[string]*ReplicaScheduleState, bool, error) {
+	initialState := make(map[string]*ReplicaScheduleState)
+	hpaControlled := false
+	hpaSelectedClusters, error := hpautil.GetHpaTargetClusterList(fedObj)
```

Comment:
I completely get your suspicion.
I had 2 options:
1. Use the cluster selector annotation itself to select the clusters to which the hpa target object should go (decided by hpa controller), overriding the user specified cluster selector (if any). In effect the hpa objects cluster selector overrides and applies to the target object also and hpa controllers decisions also apply, subsequently. This can be confusing to the user (who specified the original selector annotations on the target object) see it automagically change.
2. Let the cluster selectors on both objects work independently and use a separate annotation to communicate the hpa's intent. There can be a situation that both the selectors do not intersect, in which case no cluster will have a target object (but local hpas will exist). The fed hpa's decisions are made on observations of its local hpas, which originally get created as per the hpa cluster selector, and if the two selectors intersect, the hpa replicas would get concentrated into the intersection of clusters (which also would have target objects and thus actual status fields; currentReplicas, desiredReplicas, etc. would be updated). This result can also be confusing to the user, but does not override his/her choices. Ideally the user who creates the hpa, also created the target object and ideally should put the same cluster selector on both, but if he/she doesn't do that, the results are predictable.
I have chosen the latter (which appeared sound to me when I coded it :) ), with an intent that this is updated in docs explaining the consequences.
If you think there probably is a better way, we can have a short chat to discuss the same.